File: compiler/bench/primediv.fs
\ expected: 1229
\ Primality by trial division - count primes to 10000, run 100 times

: prime? ( n -- flag )
  dup 2 < if drop 0 exit then
  dup 2 = if drop -1 exit then
  dup 1 and 0= if drop 0 exit then
  3
  begin
    2dup dup * >= while
    2dup mod 0= if 2drop 0 exit then
    2 +
  repeat
  2drop -1 ;

: count-primes ( limit -- count )
  0 swap 2 do
    i prime? if 1+ then
  loop ;

: main ( -- )
  0
  100 0 do
    drop 10000 count-primes
  loop
  . cr ;

File: compiler/bench/primediv.c
// expected: 1229
// Primality by trial division - count primes to 10000, run 100 times

#include <stdio.h>

int prime(int n) {
    if (n < 2) return 0;
    if (n == 2) return 1;
    if ((n & 1) == 0) return 0;
    for (int i = 3; i * i <= n; i += 2) {
        if (n % i == 0) return 0;
    }
    return 1;
}

int count_primes(int limit) {
    int count = 0;
    for (int i = 2; i < limit; i++) {
        if (prime(i)) count++;
    }
    return count;
}

int main(void) {
    int result = 0;
    for (int i = 0; i < 100; i++) {
        result = count_primes(10000);
    }
    printf("%d\n", result);
    return 0;
}
